portfolio/src/main/webapp/js/map.js

```diff
@@ -13,9 +13,114 @@
+        openMarker.setMap(null);
+    }
+
+    openMarker = new google.maps.Marker({position: {lat: lat, lng: lng}, map: map});
+
+    var descriptionInputForm = createInputForm(lat, lng);
+
+    var infoWindow = new google.maps.InfoWindow({content: descriptionInputForm});
+
+    // When the user closes the open info window, remove the marker.
+    google.maps.event.addListener(infoWindow, "closeclick", () => {
+        openMarker.setMap(null);
+    });
+
+    infoWindow.open(map, openMarker);
+}
+
+//Makes a POST request to the /markers servlet to store a marker
+function submitMarker(lat, lng, description) {
+    const params = new URLSearchParams();
+    params.append("lat", lat);
+    params.append("lng", lng);
+    params.append("description", description);
+
+    fetch("/markers", {method: "POST", body: params});
+
+}
+
+//Creates a div element with a form containing a textbox and submit button for a user placed marker.
+//@return div element
+function createInputForm(lat, lng) {
+    var container = document.createElement("div");
+    var label = document.createElement("p");
+    var form = document.createElement("form");
+    var textBox = document.createElement("textarea");
+    var submitButton = document.createElement("button");
+
+    label.innerHTML = "Description:";
+    submitButton.innerHTML = "Submit";
+
+    container.classList.add("marker-form-container");
+    label.classList.add("marker-form-label");
+    form.classList.add("marker-form");
+    textBox.classList.add("marker-textbox");
+    submitButton.classList.add("marker-submit");
+    submitButton.classList.add("btn-small");
+
+    //require the textbox field
+    textBox.setAttribute("required", "");
+    textBox.setAttribute("aria-required", "true");
+    submitButton.setAttribute("type", "submit");
+
+    form.addEventListener("submit", () => {
+        submitMarker(lat, lng, textBox.value);
+        createMarker({lat: lat, lng: lng, description: textBox.value});
+        //the user has submitted the marker, so there's no longer an open marker
+        openMarker.setMap(null);
+    })
+
+    form.append(textBox);
+    form.append(submitButton);
+    container.append(label);
+    container.append(form);
+
+    return container;
+}
+
+//Loads all stored markers from the server and creates markers to display on the map.
+function loadMarkers() {
+    fetch("/markers").then(response => response.json()).then((markers) => {
+        markers.forEach((marker) => {
+            //convert data to JSON
+            createMarker(JSON.parse(marker));
+        });
+    });
+}
+
+//Creates a marker with an info window displaying the user's description.
+function createMarker(markerData) {
+    var marker = new google.maps.Marker({position: {lat: markerData.lat, lng: markerData.lng}, map: map});
+
+    var infoWindow = new google.maps.InfoWindow({content: markerData.description});
+
+    marker.addListener("click", () => {
+        infoWindow.open(map, marker);
+    });
 }
```

Comment:
Another new line here?

Reply:
I think GitHub doesn't show the new line at the end of the file if there is one, but I believe I already have new lines in these files.

Reply:
Yea, there will be a red icon if a file is missing a newline. So this one is fine.